### ISSUE TITLE
Support for Coordinate Descent

### ISSUE THREAD

**Issue by Alessio-Amaolo** [PengningChao/dolphindes#27]
It may make sense to support a memory-efficient, first order method for fine-tuning. For a massive number of constraints, the only possible option right now is GCD. At the end of GCD, fine-tuning with coordinate descent could help eek out the last bit of performance. The purpose of this would be:

1. Tighten the bound as much as possible (possibly a few percent, but maybe more depending on the GCD parameters)
2. Theoretical guarantees of convergence: GCD does great in practice, but having no guarantees of convergence means not really having a good idea if the tightest possible bound was reached
3. Provides a method of benchmarking GCD without running expensive BFGS or Newton optimization on larger problems
4. The "carry" variables described below would help keep track of future constraints not optimized in the future (such as linear zero-field constraints)

In my opinion, here are the steps for accomplishing this:

1. Write a gradient descent optimizer
2. Modify a QCQP class to have a "carry" matrix: total_A and total_S that are not modified in optimization steps, and are therefore constants between evaluations. These would correspond to the coordinates / multipliers that we are not optimizing. (Such carry variables _could_ be used with LBFGS or Newton, but for the purposes of fine-tuning it may make sense to do the cheapest possible method).
3. Optimize the rest of the coordinates by passing them to the gradient descent optimizer. Alternate which of these coordinates are the carry and optimize set to perform coordinate descent. 
4. By choosing all coordinates to be optimized, we could recover GD if calculating the objective / gradients with all values is fast enough.
